### PR TITLE
feat(mme): Send context release command on successful handover

### LINKED ISSUE #8614: [AGW][S1HO] No context release command after Handover Notify message
### Your Environment

- **Version: v1.7
- **Affected Component:** Access Gateway
- **Affected Subcomponent:** MME
- **Deployment Environment:**  bare-metal (AGW)

### Describe the Issue

The last message that is received after an S1 HO is the `Handover Notify` message. After receiving this message, the MME should send a `context release command` before the `tS1relocoverall` timer expires. Other wise the eNB sends a context release request with `ts1relocoverall-expiry` cause. 

**To Reproduce**
1. Trigger an S1HO
2. Observe call flow

**Expected behavior**
MME should send a `context release command` after receiving `HandoverNotify` message. 

**Screenshots**
image

### PR DESCRIPTION
**Pull request by talkhasib**
Signed-off-by: Tariq Al-Khasib <talkhasib@fb.com>

## Summary

In S1 handover, when MME receives a Handover Notify msg from the destination eNB, it should send a context release command to the source eNB to indicate the successful completion of the S1 handover.

## Test Plan

Tested with two physical Baicells eNBs (pcap attached)
Also fixed the corresponding unit tests
[S1_Handover.pcap.zip](https://github.com/magma/magma/files/7657622/S1_Handover.pcap.zip)

### CODE REVIEW

**Comment by sudhikan** [comment]
Thanks a ton @talkhasib for this PR :-) I am sure it will make the counters happy to report a successful handover <3

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.c [hunk 1]
```diff
@@ -3201,6 +3208,32 @@ status_code_e s1ap_mme_handle_handover_notify(
...
 
     s1ap_dump_enb(target_enb);
+
+    // Send context release command to source eNB
+    // Retrieve the association ID for the eNB that UE was connected to
+    // (i.e., Source eNB) and pull the Source eNB record from s1ap state using
+    // this association
+    if ((enb_array = hashtable_ts_get_elements(&state->enbs)) != NULL) {
+      for (idx = 0; idx < enb_array->num_elements; idx++) {
+        source_enb = (enb_description_t*) (uintptr_t) enb_array->elements[idx];
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.c]: You can get the source_enb before line 3184, by using src_ue_ref_p, as is done inside `s1ap_remove_ue`, with:

`source_enb = s1ap_state_get_enb(state, src_ue_ref_p->sctp_assoc_id);`

**Review by ssanadhya — CHANGES_REQUESTED** [review]

**Review by ssanadhya — APPROVED** [review]
Lgtm

File: lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers.cpp [hunk 2]
```diff
@@ -1053,7 +1053,10 @@ TEST_F(S1apMmeHandlersTest, HandleMmeHandover) {
 
   bool is_state_same = true;
 
-  EXPECT_CALL(*sctp_handler, sctpd_send_dl()).Times(5);
+  sctp_assoc_id_t target_assoc_id = 2;
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers.cpp:1056]: thanks for updating and fixing this

**Review by ardzoht — COMMENTED** [review]
Additional request: Can you also run `test_s1_handover.py` on the S1AP integ tests to ensure sanity?

**Comment by ardzoht** [comment]
> Additional request: Can you also run `test_s1_handover.py` on the S1AP integ tests to ensure sanity?

based on offline discussion, the mentioned integ test fails due to the changes, @talkhasib will check if this can be resolved with a fix if not we'll comment out the test and follow on the S1AP tester required changes.

File: lte/gateway/python/integ_tests/s1aptests/test_s1_handover.py [hunk 3]
```diff
@@ -63,7 +63,7 @@ def test_s1_handover(self):
         # Columns: Cell Id, Tac, EnbType, PLMN Id, PLMN length
         enb_list = [
             [1, 1, 1, "00101", 5],
-            [2, 2, 1, "00101", 5],
+            [2, 1, 1, "00101", 5],
```
Comment (ardzoht) [inline lte/gateway/python/integ_tests/s1aptests/test_s1_handover.py]: what is the context for the Tac value change?

Comment (talkhasib) [reply lte/gateway/python/integ_tests/s1aptests/test_s1_handover.py]: For some reason, running the test on my machine fails at the S1 setup stage when using different TACs.
There is no reason to use different TACs for these tests I guess

**Review by ardzoht — APPROVED** [review]
lgtm, tagging @VinashakAnkitAman for context

File: lte/gateway/python/integ_tests/s1aptests/test_s1_handover.py (hunk 3 above)
Comment (ardzoht) [reply lte/gateway/python/integ_tests/s1aptests/test_s1_handover.py]: you need to run `s1aptests/test_modify_mme_config_for_sanity.py` for this to pass, this is executed before the sanity test list to update the MME config, I'd suggest to keep this unchanged, running the above s1ap test before this one should make it work

**Review by VinashakAnkitAman — APPROVED** [review]
LGTM

**Pull request merged**